server/schedulers/hot_region.go

```diff
@@ -520,41 +630,102 @@ func (h *balanceHotRegionsScheduler) selectSrcStore(stats statistics.StoreHotPee
 // We choose a target store based on the hot region number and flow bytes of this store.
-func (h *balanceHotRegionsScheduler) selectDestStore(candidateStoreIDs []uint64, regionBytesRate float64, srcStoreID uint64, storesStat statistics.StoreHotPeersStat) (destStoreID uint64) {
+func (h *hotScheduler) selectDstStoreByHot(candidates map[uint64]struct{}, regionBytesRate float64, srcStoreID uint64, storesStat statistics.StoreHotPeersStat) (dstStoreID uint64) {
 	srcBytesRate := storesStat[srcStoreID].FutureBytesRate
-
+	srcCount := len(storesStat[srcStoreID].Stats)
 	var (
 		minBytesRate = srcBytesRate*hotRegionScheduleFactor - regionBytesRate
 		minCount     = math.MaxInt32
 	)
-	for _, storeID := range candidateStoreIDs {
+	for storeID := range candidates {
 		if s, ok := storesStat[storeID]; ok {
-			count, dstBytesRate := len(s.Stats), math.Max(s.StoreBytesRate, s.FutureBytesRate)
-			if minBytesRate > dstBytesRate || (minBytesRate == dstBytesRate && minCount > count) {
-				minCount = count
-				minBytesRate = dstBytesRate
-				destStoreID = storeID
+			dstCount, dstBytesRate := len(s.Stats), math.Max(s.StoreBytesRate, s.FutureBytesRate)
+			if pickByFlowBytes {
+				if minBytesRate > dstBytesRate || (minBytesRate == dstBytesRate && minCount > dstCount) {
+					minCount = dstCount
+					minBytesRate = dstBytesRate
+					dstStoreID = storeID
+				}
+			} else {
+				if srcCount < dstCount+2 { // ensure srcCount >= dstCount after the operation.
+					continue
+				}
+				if minCount > dstCount || (minCount == dstCount && minBytesRate > dstBytesRate) {
+					dstStoreID = storeID
+					minBytesRate = dstBytesRate
+					minCount = dstCount
+				}
 			}
 		} else {
-			destStoreID = storeID
-			return
+			return storeID
 		}
 	}
-	return
+	return dstStoreID
 }
 
-func (h *balanceHotRegionsScheduler) adjustBalanceLimit(storeID uint64, storesStat statistics.StoreHotPeersStat) uint64 {
+func (h *hotScheduler) selectSrcStoreByScore(stats statistics.StoreHotPeersStat, balanceType BalanceType) uint64 {
+	var storesScore *ScoreInfos
+	if balanceType == hotRead {
+		storesScore = h.readScores
+	} else {
+		storesScore = h.writeScores
+	}
+	if storesScore.Variation() > minScoreVariationLimit {
+		maxScore := storesScore.Max()                 // with sort
+		for i := storesScore.Len() - 1; i >= 0; i-- { // select from large to small
+			scoreInfo := storesScore.scoreInfos[i]
+			if scoreInfo.GetScore() >= maxScore*hotRegionScheduleFactor {
+				storeID := scoreInfo.GetStoreID()
+				if stat, ok := stats[storeID]; ok && len(stat.Stats) > 1 {
+					return storeID
+				}
+			}
+		}
+	}
+	return 0
+}
+
+func (h *hotScheduler) selectDstStoreByScore(candidates map[uint64]struct{}, regionBytesRate float64, balanceType BalanceType) uint64 {
+	var storesScore *ScoreInfos
+	if balanceType == hotRead {
+		storesScore = h.readScores
+	} else {
+		storesScore = h.writeScores
+	}
+	if storesScore.Variation() > minScoreVariationLimit {
```

Comment:
Simply compare the "variation" is not working well:
![image](https://user-images.githubusercontent.com/17215435/71892902-2d618a80-3185-11ea-946f-499f3b7d5b5b.png)
I think we can remove this and find a better way to prevent overschedule.

Reply:
Well

Reply:
I add `checkVariation` to ensure that variation will become smaller.PTAL